adapters/base/src/main/java/org/apache/cassandra/sidecar/adapters/base/TokenRangeReplicaProvider.java

```diff
@@ -0,0 +1,249 @@
+{
+    private final JmxClient jmxClient;
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(TokenRangeReplicaProvider.class);
+
+    public TokenRangeReplicaProvider(JmxClient jmxClient)
+    {
+        this.jmxClient = jmxClient;
+    }
+
+    public TokenRangeReplicasResponse tokenRangeReplicas(String keyspace, Partitioner partitioner)
+    {
+        Objects.requireNonNull(keyspace, "keyspace must be non-null");
+
+        StorageJmxOperations storage = jmxClient.proxy(StorageJmxOperations.class, STORAGE_SERVICE_OBJ_NAME);
+
+        // Retrieve map of primary token ranges to endpoints that describe the ring topology
+        Map<List<String>, List<String>> rangeToEndpointMappings = storage.getRangeToEndpointWithPortMap(keyspace);
+        // Pending ranges include bootstrap tokens and leaving endpoints as represented in the Cassandra TokenMetadata
+        Map<List<String>, List<String>> pendingRangeMappings = storage.getPendingRangeToEndpointWithPortMap(keyspace);
+
+        Set<String> replicaSet = Stream.concat(rangeToEndpointMappings.values().stream().flatMap(List::stream),
+                                               pendingRangeMappings.values().stream().flatMap(List::stream))
+                                       .collect(Collectors.toSet());
+
+        Map<String, String> hostToDatacenter = groupHostsByDatacenter(replicaSet);
+
+        // Retrieve map of all token ranges (pending & primary) to endpoints
+        List<TokenRangeReplicasResponse.ReplicaInfo> writeReplicas =
+        writeReplicasFromPendingRanges(rangeToEndpointMappings,
+                                       pendingRangeMappings,
+                                       hostToDatacenter,
+                                       partitioner,
+                                       keyspace);
+
+        Map<String, String> replicaToStateMap = replicaToStateMap(replicaSet, storage);
+
+        return new TokenRangeReplicasResponse(
+        replicaToStateMap,
+        writeReplicas,
+        mappingsToUnwrappedReplicaSet(rangeToEndpointMappings, hostToDatacenter, partitioner));
+    }
+
+    private Map<String, String> replicaToStateMap(Set<String> replicaSet, StorageJmxOperations storage)
+    {
+        List<String> joiningNodes = storage.getJoiningNodesWithPort();
+        List<String> leavingNodes = storage.getLeavingNodesWithPort();
+        List<String> movingNodes = storage.getMovingNodesWithPort();
+
+        String rawGossipInfo = getRawGossipInfo();
+        GossipInfoResponse gossipInfo = GossipInfoParser.parse(rawGossipInfo);
+
+        StateWithReplacement state = new StateWithReplacement(joiningNodes, leavingNodes, movingNodes, gossipInfo);
+
+        return replicaSet.stream()
+                         .collect(Collectors.toMap(Function.identity(), state::of));
+    }
+
+    private String getRawGossipInfo()
+    {
+        return jmxClient.proxy(ClusterMembershipJmxOperations.class, FAILURE_DETECTOR_OBJ_NAME)
```

Comment:
Similar comment to above, also make sure the initialization methods are protected so they can be overridden 
```suggestion
        return initializeClusterMembershipOps()
```

Reply:
Same (make things protected) goes for any members that need to be accessible by the derived class (or provide a protected getter if you feel that's necessary, but I think it's mostly the `private final JmxClient jmxClient`, which would be OK to be protected).

Reply:
Will make the change similar to how RingHandler does this for consistency